File: examples/Ruby/lbbroker.rb
# Load-balancing broker
# Clients and workers are shown here in-process

require 'rubygems'
require 'ffi-rzmq'

CLIENT_SIZE = 10
WORKER_SIZE = 3

def client_task(identity)
  context = ZMQ::Context.new
  client = context.socket ZMQ::REQ
  client.identity = identity
  client.connect "ipc://frontend.ipc"

  client.send_string "HELLO"
  client.recv_string reply = ""

  puts "#{identity}: #{reply}"

  client.close
  context.destroy
end

def worker_task(identity)
  context = ZMQ::Context.new
  worker = context.socket ZMQ::REQ
  worker.identity = identity
  worker.connect "ipc://backend.ipc"

  worker.send_string "READY"

  loop do
    worker.recv_string client = ""
    worker.recv_string empty = ""
    worker.recv_string request = ""

    puts "#{identity}: #{request} from #{client}"

    worker.send_strings [client, empty, "OK from #{identity}"]
  end

  worker.close
  context.destroy
end

def main_task
  context = ZMQ::Context.new
  frontend = context.socket ZMQ::ROUTER
  backend = context.socket ZMQ::ROUTER

  frontend.bind "ipc://frontend.ipc"
  backend.bind "ipc://backend.ipc"

  CLIENT_SIZE.times do |client_id|
    Thread.new { client_task "CLIENT-#{client_id}" }
  end

  WORKER_SIZE.times do |worker_id|
    Thread.new { worker_task "WORKER-#{worker_id}" }
  end

  available_workers = []
  poller = ZMQ::Poller.new
  poller.register_readable backend
  poller.register_readable frontend

  # The poller will continuously poll the backend and will poll the
  # frontend when there is at least one worker available.

  while poller.poll > 0
    poller.readables.each do |readable|
      if readable === backend
        backend.recv_string worker = ""
        backend.recv_string empty = ""
        backend.recv_strings reply = []

        frontend.send_strings reply unless reply[0] == "READY"

        # Add this worker to the list of available workers
        available_workers << worker
      elsif readable === frontend && available_workers.any?
        # Read the request from the client and forward it to the LRU worker
        frontend.recv_strings request = []
        backend.send_strings [available_workers.shift, ""] + request
      end
    end
  end

  frontend.close
  backend.close
  context.destroy
end

main_task

File: examples/Python/lbbroker.py
"""
Load-balancing broker

Clients and workers are shown here in-process.

Author: Brandon Carpenter (hashstat) <brandon(dot)carpenter(at)pnnl(dot)gov>
"""

from __future__ import print_function

import multiprocessing

import zmq


NBR_CLIENTS = 10
NBR_WORKERS = 3


def client_task(ident):
    """Basic request-reply client using REQ socket."""
    socket = zmq.Context().socket(zmq.REQ)
    socket.identity = u"Client-{}".format(ident).encode("ascii")
    socket.connect("ipc://frontend.ipc")

    # Send request, get reply
    socket.send(b"HELLO")
    reply = socket.recv()
    print("{}: {}".format(socket.identity.decode("ascii"),
                          reply.decode("ascii")))


def worker_task(ident):
    """Worker task, using a REQ socket to do load-balancing."""
    socket = zmq.Context().socket(zmq.REQ)
    socket.identity = u"Worker-{}".format(ident).encode("ascii")
    socket.connect("ipc://backend.ipc")

    # Tell broker we're ready for work
    socket.send(b"READY")

    while True:
        address, empty, request = socket.recv_multipart()
        print("{}: {}".format(socket.identity.decode("ascii"),
                              request.decode("ascii")))
        socket.send_multipart([address, b"", b"OK"])


def main():
    """Load balancer main loop."""
    # Prepare context and sockets
    context = zmq.Context.instance()
    frontend = context.socket(zmq.ROUTER)
    frontend.bind("ipc://frontend.ipc")
    backend = context.socket(zmq.ROUTER)
    backend.bind("ipc://backend.ipc")

    # Start background tasks
    def start(task, *args):
        process = multiprocessing.Process(target=task, args=args)
        process.daemon = True
        process.start()
    for i in range(NBR_CLIENTS):
        start(client_task, i)
    for i in range(NBR_WORKERS):
        start(worker_task, i)

    # Initialize main loop state
    count = NBR_CLIENTS
    workers = []
    poller = zmq.Poller()
    # Only poll for requests from backend until workers are available
    poller.register(backend, zmq.POLLIN)

    while True:
        sockets = dict(poller.poll())

        if backend in sockets:
            # Handle worker activity on the backend
            request = backend.recv_multipart()
            worker, empty, client = request[:3]
            if not workers:
                # Poll for clients now that a worker is available
                poller.register(frontend, zmq.POLLIN)
            workers.append(worker)
            if client != b"READY" and len(request) > 3:
                # If client reply, send rest back to frontend
                empty, reply = request[3:]
                frontend.send_multipart([client, b"", reply])
                count -= 1
                if not count:
                    break

        if frontend in sockets:
            # Get next client request, route to last-used worker
            client, empty, request = frontend.recv_multipart()
            worker = workers.pop(0)
            backend.send_multipart([worker, b"", client, b"", request])
            if not workers:
                # Don't poll clients if no workers are available
                poller.unregister(frontend)

    # Clean up
    backend.close()
    frontend.close()
    context.term()


if __name__ == "__main__":
    main()
